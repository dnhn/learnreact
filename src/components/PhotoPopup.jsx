import React, { Component } from 'react';
import { connect } from 'react-redux';
import { closePhoto } from '../redux/actions';

import './PhotoPopup.scss';

export class PhotoPopup extends Component {
  _timeOut = false;

  constructor(props) {
    super(props);
    this.state = {
      show: false,
    };
  }

  componentDidMount() {
    this._timeOut = setTimeout(() => {
      this.setState({
        show: true,
      });
      this.closeBtn.focus();
    }, 0);
  }

  componentWillUnmount() {
    clearTimeout(this._timeOut);
  }

  render() {
    const {
      selectedPhoto: {
        description,
        urls,
        user,
      },
      closePhoto,
    } = this.props;
    const { show } = this.state;

    return (
      <div className={`photo-popup ${show ? 'show' : ''}`}>
        <div
          className="photo-popup__photo"
          style={{backgroundImage: `url(${urls.regular})`}}
        />
        <p className="photo-popup__caption">
          {description ? <em>“{description}”</em> : ''}
          {description ? ' — ' : ''}
          by&nbsp;
          <a
            href={user.links.html}
            target="_blank"
            rel="noopener noreferrer"
            className="photo-popup__author-link"
          >
            {`${user.first_name || ''} ${user.last_name || ''}`}
          </a>
        </p>
        <button
          ref={ref => this.closeBtn = ref}
          type="button"
          onClick={closePhoto}
          className="photo-popup__close"
        >
          Close
        </button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  selectedPhoto: state.selectedPhoto,
});

const mapDispatchToProps = dispatch => ({
  closePhoto: () => dispatch(closePhoto()),
});

export default connect(mapStateToProps, mapDispatchToProps)(PhotoPopup);
