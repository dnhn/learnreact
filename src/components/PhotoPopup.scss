.photo-popup {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  height: 100%;
  padding: 3%;
  background-color: rgba(0, 0, 0, .75);
  color: white;
  display: flex;
  flex-direction: column;
  align-items: center;
  transition: opacity .5s ease;

  &:not(.show) {
    opacity: 0;
  }
}

.photo-popup__photo {
  width: 100%;
  height: 100%;
  background-size: contain;
  background-position: center;
  background-repeat: no-repeat;
}

.photo-popup__caption {
  width: 95%;
  text-align: center;
}

.photo-popup__author-link {
  color: #ccc;

  &:hover,
  &:focus,
  &:active {
    color: white;
  }
}

.photo-popup__close {
  position: fixed;
  top: 24px;
  left: 24px;
  padding: 2%;
  background-color: white;
  border-radius: 2px;
  box-shadow: 1px 1px 0 black;
}

@media screen and (max-width: 640px) {
  .photo-popup__close {
    opacity: .75;
  }
}

@media screen and (min-width: 641px) {
  .photo-popup__caption {
    width: 75%;
  }

  .photo-popup__close {
    top: 16px;
    left: 16px;
    padding: 1%;
  }
}
