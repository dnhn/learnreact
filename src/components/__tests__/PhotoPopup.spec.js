import React from 'react';
import { fireEvent, render } from "@testing-library/react";

import { PhotoPopup } from '../PhotoPopup';

let mockStore;

beforeAll(() => {
  jest.useFakeTimers();
});

beforeEach(() => {
  mockStore = {
    selectedPhoto: {
      description: 'description',
      urls: { regular: '#background' },
      user: {
        links: { html: '#link' },
        first_name: 'First',
        last_name: 'Last',
      },
    },
    closePhoto: jest.fn(),
  };
});

afterAll(() => {
  jest.useRealTimers();
});

describe('<PhotoPopup />', () => {
  it('renders without crashing', () => {
    const { asFragment } = render(<PhotoPopup {...mockStore} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('has no description', () => {
    mockStore.selectedPhoto.description = '';
    const { container } = render(<PhotoPopup {...mockStore} />);
    expect(container.getElementsByTagName('em').length)
      .toBe(0);
  });

  it('has no first name', () => {
    mockStore.selectedPhoto.user.first_name = '';
    const { container } = render(<PhotoPopup {...mockStore} />);
    expect(
      container
        .getElementsByClassName('photo-popup__author-link')[0]
        .textContent
        .trim()
    )
      .toBe(mockStore.selectedPhoto.user.last_name);
  });

  it('has no last name', () => {
    mockStore.selectedPhoto.user.last_name = '';
    const { container } = render(<PhotoPopup {...mockStore} />);
    expect(
      container
        .getElementsByClassName('photo-popup__author-link')[0]
        .textContent
        .trim()
    )
      .toBe(mockStore.selectedPhoto.user.first_name);
  });

  it('closes on click', () => {
    const { getByText } = render(<PhotoPopup {...mockStore} />);
    fireEvent.click(getByText('Close'));
    expect(mockStore.closePhoto).toHaveBeenCalled();
  });
});
