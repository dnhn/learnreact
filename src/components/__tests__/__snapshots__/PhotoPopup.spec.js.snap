// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoPopup /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="photo-popup "
  >
    <div
      class="photo-popup__photo"
      style="background-image: url(#background);"
    />
    <p
      class="photo-popup__caption"
    >
      <em>
        “description”
      </em>
       — by 
      <a
        class="photo-popup__author-link"
        href="#link"
        rel="noopener noreferrer"
        target="_blank"
      >
        First Last
      </a>
    </p>
    <button
      class="photo-popup__close"
      type="button"
    >
      Close
    </button>
  </div>
</DocumentFragment>
`;
